external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/producer/InternalKafkaProducerPool.scala

```diff
@@ -0,0 +1,205 @@
+ */
+private[producer] class InternalKafkaProducerPool(
+    executorService: ScheduledExecutorService,
+    val clock: Clock,
+    conf: SparkConf) extends Logging {
+  import InternalKafkaProducerPool._
+
+  def this(sparkConf: SparkConf) = {
+    this(ThreadUtils.newDaemonSingleThreadScheduledExecutor(
+      "kafka-producer-cache-evictor"), new SystemClock, sparkConf)
+  }
+
+  /** exposed for testing */
+  private[producer] val cacheExpireTimeout: Long = conf.get(PRODUCER_CACHE_TIMEOUT)
+
+  private val evictorThreadRunIntervalMillis = conf.get(PRODUCER_CACHE_EVICTOR_THREAD_RUN_INTERVAL)
+
+  @GuardedBy("this")
+  private val cache = new mutable.HashMap[CacheKey, CachedProducerEntry]
+
+  private def startEvictorThread(): Option[ScheduledFuture[_]] = {
+    if (evictorThreadRunIntervalMillis > 0) {
+      val future = executorService.scheduleAtFixedRate(() => {
+        Utils.tryLogNonFatalError(evictExpired())
+      }, 0, evictorThreadRunIntervalMillis, TimeUnit.MILLISECONDS)
+      Some(future)
+    } else {
+      None
+    }
+  }
+
+  private var scheduled = startEvictorThread()
+
+  /**
+   * Get a cached KafkaProducer for a given configuration. If matching KafkaProducer doesn't
+   * exist, a new KafkaProducer will be created. KafkaProducer is thread safe, it is best to keep
+   * one instance per specified kafkaParams.
+   */
+  private[producer] def acquire(kafkaParams: ju.Map[String, Object]): CachedKafkaProducer = {
+    val updatedKafkaProducerConfiguration =
+      KafkaConfigUpdater("executor", kafkaParams.asScala.toMap)
+        .setAuthenticationConfigIfNeeded()
+        .build()
+    val paramsSeq: Seq[(String, Object)] = paramsToSeq(updatedKafkaProducerConfiguration)
+    synchronized {
+      val entry = cache.getOrElseUpdate(paramsSeq, {
+        val producer = createKafkaProducer(paramsSeq)
+        val cachedProducer = new CachedKafkaProducer(paramsSeq, producer)
+        new CachedProducerEntry(cachedProducer, clock, cacheExpireTimeout)
+      })
+      entry.handleBorrowed()
+      entry.producer
+    }
+  }
+
+  private[producer] def release(producer: CachedKafkaProducer): Unit = {
+    def closeProducerNotInCache(producer: CachedKafkaProducer): Unit = {
+      logWarning(s"Released producer ${producer.id} is not a member of the cache. Closing.")
+      producer.close()
+    }
+
+    synchronized {
+      cache.get(producer.cacheKey) match {
+        case Some(entry) if entry.producer.id == producer.id => entry.handleReturned()
+        case _ => closeProducerNotInCache(producer)
+      }
+    }
+  }
+
+  private[producer] def shutdown(): Unit = {
+    ThreadUtils.shutdown(executorService)
+  }
+
+  /** exposed for testing. */
+  private[producer] def reset(): Unit = synchronized {
+    scheduled.foreach(_.cancel(true))
+    cache.foreach { case (k, v) =>
+      cache.remove(k)
+      v.producer.close()
+    }
+    scheduled = startEvictorThread()
```

Comment:
Hmm, not sure why you need to restart this task?

Reply:
Because `reset` is to ensure the state of producer pool is back to first initialization; so canceling the evict task to make sure evict task is not running, closing and clearing all cached producers, restarting the evict task.

It's ideal to simply recreate pool, but in tests we would like to reset pool in object instead of class instance. It's only used from testing.

Reply:
Well, if you want to be absolutely sure that everything is being reset then, you need to use `cancel(false)`, and wait for any pending run of the task to finish. Otherwise you may interrupt the task in the middle of closing an expired producer. Or maybe tests don't run into that because [insert reason here].

It just seems safer to me to not bother messing with the task. Leaving it alone will not break anything, and will reset the pool just as well.

Reply:
> It just seems safer to me to not bother messing with the task. Leaving it alone will not break anything, and will reset the pool just as well.

Agreed. Not touching the evict task while resetting seems to also work. Will change.